### PR TITLE
[clang] Sequence C++20 Parenthesized List Init

### LINKED ISSUE #83474: [clang] False positive "-Wunsequenced" for C++20 parenthesized initializer lists
Clang 17.0.6 and the latest git/main branch will issue "-Wunsequenced" for the following code sample,

```c++
int val = 0;
struct S { int a; int b; };

int paren_list_init[](val++, val++);
int direct_list_init[]{val++, val++};

S paren_list_init_s(val++, val++);
S direct_list_init_s{val++, val++};
```

for only the C++20 parenthesized initializer lists, even though both are sequenced operations.

```
$ guix shell clang -- clang -c -o /dev/null -Wunsequenced -std=c++20 test.cpp 
test.cpp:4:26: warning: multiple unsequenced modifications to 'val' [-Wunsequenced]
    4 | int paren_list_init[](val++, val++);
      |                          ^      ~~
test.cpp:7:24: warning: multiple unsequenced modifications to 'val' [-Wunsequenced]
    7 | S paren_list_init_s(val++, val++);
      |                        ^      ~~
2 warnings generated.

```

I believe this to be a false positive because the C++20 parenthesized initializer lists added in [p0960r3](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0960r3.html) explicitly state that the values and sides effects are explicitly sequenced, and provides the ordering. See C++20 [decl.init]p16.5 and [decl.init]p16.6.2.2, or the latest draft [here](https://www.eel.is/c++draft/dcl.init#general-16.5) and [here](https://www.eel.is/c++draft/dcl.init#general-16.6.2.2)

### PR DESCRIPTION
**Pull request by vapdrs**
Parenthesized list intializers are sequenced operations, see C++20 [decl.init]p16.5 and [decl.init]p16.6.2.2 for more details.

Fixes #83474

### CODE REVIEW

File: clang/lib/Sema/SemaChecking.cpp
Comment (zygoloid) [inline clang/lib/Sema/SemaChecking.cpp:17632]: I think you mean `dcl.init.general`, not `decl.init` here, and the wording is in p16 not p17 in C++20 and the current working paper.
```suggestion
    // [dcl.init.general]p16.5 and [dcl.init.general]p16.6.2.2.
```

File: clang/lib/Sema/SemaChecking.cpp
Comment (zygoloid) [inline clang/lib/Sema/SemaChecking.cpp:17647]: Is it feasible to factor out the common code between this and `VisitInitListExpr`? This looks nearly identical.

**Review by zygoloid — COMMENTED** [review]
Thanks!

[I'm not sure when I'll have time to circle back to this, so I'd be happy for someone else to finish the review and approve.]

File: clang/lib/Sema/SemaChecking.cpp
Comment (shafik) [inline clang/lib/Sema/SemaChecking.cpp:17618]: I took me a bit of checking to convince myself this was doing the right thing. It might be nice to refactor `CXXConstructExpr` to have an a member that does the same as `ILE->inits()` and returns an `ArrayRef`. It looks like do similar things to create an `ArrayRef` in other places as well but probably should be a second PR.

Maybe change this to `llvm::ArrayRef(CCE->getArgs(), CCE->getNumArgs())` to make it more explicit for now.

**Review by shafik — COMMENTED** [review]
LGTM after addressing comment.

**Review by shafik — APPROVED** [review]

**Review by shafik — APPROVED** [review]

**Comment by cor3ntin** [comment]
@vapdrs Do you need us to merge that for you? Thanks

**Comment by vapdrs** [comment]
> @vapdrs Do you need us to merge that for you? Thanks

Yes, I do not have write access to the repository. Thank you!

**Pull request merged**
